Messy.py

```diff
@@ -0,0 +1,252 @@
+        verbose: bool = False,
+    ) -> None:
+        """
+        Copy files to the remote server
+        """
+        if remote_folder == "~":
+            remote_folder = os.path.expanduser("~")
+
+        local_folder_dirname = os.path.dirname(local_folder)
+        local_folder_basename = os.path.basename(local_folder)
+
+        if verbose:
+            print(f"local_folder_dirname: {local_folder_dirname}")
+            print(f"local_folder_basename: {local_folder_basename}")
+
+        for dirpath, dirnames, filenames in os.walk(local_folder):
+            file_dirpath = dirpath[len(local_folder) + 1 :]
+            remote_path = os.path.join(
+                remote_folder, local_folder_basename, file_dirpath
+            )
+            if verbose:
+                print("\n" + "-" * 20 + "\n")
+                print(f"dirpath: {dirpath}")
+                print(f"dirnames: {dirnames}")
+                print(f"remote_path: {remote_path}")
+                print(f"file_dirpath: {file_dirpath}")
+
+            try:
+                self.sftp_client.listdir(remote_path)
+            except IOError:
+                self.sftp_client.mkdir(remote_path)
+
+            for filename in filenames:
+                from_path = os.path.join(dirpath, filename)
+                to_path = os.path.join(remote_path, filename)
+                self.sftp_client.put(from_path, to_path)
+                if verbose:
+                    print(
+                        f"\ncopying: {filename}\n  from: {from_path}\n  to: {to_path}"
+                    )
+
+    def submit_job(self, job_file: str) -> None:
+        """
+        Submit job to SLURM cluster
+        """
+        stdin, stdout, stderr = self.ssh_client.exec_command(
+            f"sbatch {job_file}"
+        )
+        job_id = int(stdout.read().decode().split()[-1])
+        print(f"Submitted job with id: {job_id}")
+        return job_id, stdin, stdout, stderr
+
+    def build_apptainer_image(self, netlogo_version: str) -> bool:
+        """Builds a netlogo apptainer image"""
+        docker_image = f"comses/netlogo:{netlogo_version}"
+        apptainer_file = f"netlogo_{netlogo_version}.sif"
+        build_command = f"apptainer pull docker://{docker_image}"
+
+        process = subprocess.Popen(build_command, shell=True)
+        process.wait()
+
+        # TODO: handle process status mor ecarefully: what happens if the file exists?
+        if os.path.isfile(apptainer_file):
+            print(f"Generated {apptainer_file}")
+            # print(process.stdout)
+            return True
+        else:
+            print("Image generation has failed.")
+            print(process.stderr)
+            return False
+
+    def check_command_exists(self, command: str) -> bool:
+        """Check whether `command` is on PATH and marked as executable."""
+        return which(command) is not None
+
+    def check_required_tools(self, command_list: list[str]) -> bool:
```

Comment:
I think you also need a function to load modules (like apptainer) on slurm clusters using the module command.

Reply:
This should be another method. I created an issue about this.